sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Processor/EventProcessorClientTests.cs

```diff
@@ -2072,17 +2008,97 @@ private static EventHubConnectionOptions GetConnectionOptionsSample(EventProcess
         {
             for (var index = 0; index < eventCount; ++index)
             {
-                if (cancellationToken.IsCancellationRequested)
-                { break; }
-                await Task.Delay(25).ConfigureAwait(false);
+                await Task.Delay(5);
+                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
                 yield return new PartitionEvent();
             }
 
-            yield break;
+            await Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///   Allows for injecting a consumer client to use as the source of events to be processed.
+        /// </summary>
+        ///
+        private class InjectableEventSourceProcessorMock : EventProcessorClient
+        {
+            private Func<ProcessErrorEventArgs, Task> _processErrorAsync;
+            private EventHubConsumerClient _consumer;
+            private bool _stopCalled;
+
+            public new event Func<ProcessErrorEventArgs, Task> ProcessErrorAsync
+            {
+                add
+                {
+                    Argument.AssertNotNull(value, nameof(ProcessErrorAsync));
+
+                    if (IsRunning)
+                    {
+                        throw new InvalidOperationException("Can't manipulate event handlers when running.");
+                    }
+
+                    _processErrorAsync = value;
+                }
+
+                remove
+                {
+                    Argument.AssertNotNull(value, nameof(ProcessErrorAsync));
+
+                    if (IsRunning)
+                    {
+                        throw new InvalidOperationException("Can't manipulate event handlers when running.");
+                    }
+
+                    _processErrorAsync = default;
+                }
+            }
+
+            public InjectableEventSourceProcessorMock(PartitionManager storageManager,
```

Comment:
I like your approach, but there's still one issue. Even though we are not triggering the error handler anymore, we are still triggering the stop handler with the wrong close reason.

https://github.com/Azure/azure-sdk-for-net/blob/e50c09986781a012e2a8d978bc66db35d97967ff/sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs#L1283-L1303

As you can see at the end of the snippet of code, when an exception occurs in the processing task, we update the reason to `OwnershipLost` (even though a shutdown has been requested) so the user knows they cannot checkpoint.

Do you think that overriding `RunPartitionProcessingAsync` would be a better a idea? It's already marked as `internal virtual`. We could just call the base, catch any exception, filter it and bubble it up to the caller.

https://github.com/Azure/azure-sdk-for-net/blob/e50c09986781a012e2a8d978bc66db35d97967ff/sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs#L826

Reply:
ooohhhhh.....  that's a much better idea, and much less hacky.   Giving that a try.